shared/src/balancer/pool_fetching.rs

```diff
@@ -0,0 +1,169 @@
+}
+
+impl WeightedPool {
+    fn new(pool_data: RegisteredPool, balances: Vec<U256>, weights: Vec<U256>) -> Self {
+        let mut reserves = HashMap::new();
+        // We expect the weight and token indices are aligned with balances returned from EVM query.
+        // If necessary we would also pass the tokens along with the query result,
+        // use them and fetch the weights from the registry by token address.
+        for (i, balance) in balances.into_iter().enumerate() {
+            reserves.insert(
+                pool_data.tokens[i],
+                PoolTokenState {
+                    balance,
+                    weight: weights[i],
+                },
+            );
+        }
+        WeightedPool {
+            pool_id: pool_data.pool_id,
+            pool_address: pool_data.pool_address,
+            specialization: pool_data.specialization,
+            reserves,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+pub trait WeightedPoolFetching: Send + Sync {
+    async fn fetch(
+        &self,
+        token_pairs: HashSet<TokenPair>,
+        at_block: Block,
+    ) -> Result<Vec<WeightedPool>>;
+}
+
+pub struct BalancerPoolFetcher {
+    pool_data: PoolRegistry,
+    vault: BalancerV2Vault,
+    web3: Web3,
+}
+
+#[async_trait::async_trait]
+impl WeightedPoolFetching for BalancerPoolFetcher {
+    async fn fetch(
+        &self,
+        token_pairs: HashSet<TokenPair>,
+        at_block: Block,
+    ) -> Result<Vec<WeightedPool>> {
+        let mut batch = CallBatch::new(self.web3.transport());
+        let block = BlockId::Number(at_block.into());
+        let futures = token_pairs
+            .into_iter()
+            .flat_map(|pair| self.pool_data.pools_containing_pair(pair))
+            .unique_by(|pool| pool.pool_id)
+            .map(|weighted_pool| {
+                let pool_contract =
+                    BalancerV2WeightedPool::at(&self.web3, weighted_pool.pool_address);
+                let weights = pool_contract
+                    .methods()
+                    .get_normalized_weights()
+                    .batch_call(&mut batch);
+                let reserves = self
+                    .vault
+                    .get_pool_tokens(Bytes(weighted_pool.pool_id.0))
+                    .block(block)
+                    .batch_call(&mut batch);
+                async move {
+                    // Clippy is wrong about this being eval order dependent.
+                    #[allow(clippy::eval_order_dependence)]
+                    FetchedWeightedPool {
+                        pool_data: weighted_pool,
+                        reserves: reserves.await,
+                        weights: weights.await,
+                    }
+                }
+            })
+            .collect::<Vec<_>>();
+        batch.execute_all(MAX_BATCH_SIZE).await;
+
+        let mut results = Vec::new();
+        for future in futures {
+            results.push(future.await);
```

Comment:
nit: maybe comment that given the batch is executed these awaits should resolve immediately? Otherwise a future maintainer might wonder why they are not parallelised.

Reply:
Also will be addressed in #698